sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -1268,4 +1269,16 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       assert(getNumSortsInQuery(query5) == 1)
     }
   }
+
+  test("SPARK-25482: Reuse same Subquery in order to execute it only once") {
+    withTempView("t1", "t2") {
+      sql("create temporary view t1(a int) using parquet")
+      sql("create temporary view t2(b int) using parquet")
+      val plan = sql("select * from t2 where b > (select max(a) from t1)")
```

Comment:
sorry it has been a long time and I don't quite remember the context.

What was the problem we are trying to fix? This test looks nothing related to subquery reuse.

Reply:
Sure, please can you check the PR description? I think the context is quite well explained there.

Anyway, as a quick summary: in this case `b > (select max(a) from t1)` is pushed down as a datasource filter. So we have 2 instances of `b > (select max(a) from t1)` and the result is not reused. It is not reused because the copied plan satisfies `==`, so even if `ReuseSubquery` replaces it, then the change is ignored.

Reply:
Do we only have a problem when we have subquery in data source filter? If that's the case, I would suggest not pushdown subquery filter into data source.

Reply:
we have this problem if we `copy` the same subquery. I can't think of any other case than filter push-down, but I may be wrong.

Forbidding to push down these filter may cause a perf regression, I am not sure it is the right solution.

Reply:
is there any data source can support subquery filter? for data source v1/v2, the public `Filter` API does not support subquery. For file source, they don't support subquery filter either.

Reply:
I see your point. I'd argue that scalar subqueries like this may be used for Filter pushdown: since the subquery is evaluated before the plan execution, their result _may_ be used in the filter push down. With the current codebase you're 100% right and I can follow your comment/suggestion. But wouldn't it be worth to let them be used (and do that in another PR)?

Reply:
I think you are right about it, but it also means the data source scan must wait until the subquery is finished. We need to make tradeoffs carefully.

I'd suggest we open a new ticket about scalar subquery filter pushdown to data source, and forbid it here.

Reply:
> The subquery should be executed anyway sooner or later, right?

Yes, but we could execute scan and subquery at the same time (2 spark jobs running together), instead of executing them serialized.

Reply:
> we could execute scan and subquery at the same time (

is this really possible? My understanding is that subqueries are executed before the plan they belong to (in `SparkPlan.executeQuery`). So my understanding is that when a subquery is running, the rest of the query is not.

Reply:
ah sorry I misread the code. Unless the subquery is rewritten into join, we must wait for all subqueries to be finished before executing the plan.

We can rewrite scalar subquery in data source filters into literal, to make it work with the filter pushdown API.

Reply:
yes, exactly. That's what I meant. Shall I revert the changes to the previous PR? And in the scope of the new JIRA do the rewriting to literals? Thanks.